src/tasks/WasmAppBuilder/coreclr/SignatureMapper.cs

```diff
@@ -0,0 +1,144 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using WasmAppBuilder;
+
+internal static class SignatureMapper
+{
+    internal static char? TypeToChar(Type t, LogAdapter log, out bool isByRefStruct, int depth = 0)
+    {
+        isByRefStruct = false;
+
+        if (depth > 5) {
+            log.Warning("WASM0064", $"Unbounded recursion detected through parameter type '{t.Name}'");
+            return null;
+        }
+
+        char? c = null;
+        if (t.Namespace == "System") {
+            c = t.Name switch
+            {
+                nameof(String) => 'i',
+                nameof(Boolean) => 'i',
+                nameof(Char) => 'i',
+                nameof(SByte) => 'i',
+                nameof(Byte) => 'i',
+                nameof(Int16) => 'i',
+                nameof(UInt16) => 'i',
+                nameof(Int32) => 'i',
+                nameof(UInt32) => 'i',
+                nameof(Int64) => 'l',
+                nameof(UInt64) => 'l',
+                nameof(Single) => 'f',
+                nameof(Double) => 'd',
+                // FIXME: These will need to be L for wasm64
+                nameof(IntPtr) => 'i',
+                nameof(UIntPtr) => 'i',
+                "Void" => 'v',
+                _ => null
+            };
+        }
+
+        if (c != null)
+            return c;
+
+        // FIXME: Most of these need to be L for wasm64
+        if (t.IsArray)
+            c = 'i';
+        else if (t.IsByRef)
+            c = 'i';
+        else if (typeof(Delegate).IsAssignableFrom(t))
+            // FIXME: Should we narrow this to only certain types of delegates?
+            c = 'i';
+        else if (t.IsClass)
+            c = 'i';
+        else if (t.IsInterface)
+            c = 'i';
+        else if (t.IsEnum) {
+            Type underlyingType = t.GetEnumUnderlyingType();
+            c = TypeToChar(underlyingType, log, out _, ++depth);
+        } else if (t.IsPointer)
+            c = 'i';
+        else if (PInvokeTableGenerator.IsFunctionPointer(t))
+            c = 'i';
+        else if (t.IsValueType)
+        {
+            var fields = t.GetFields(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+            if (fields.Length == 1)
+            {
+                Type fieldType = fields[0].FieldType;
+                return TypeToChar(fieldType, log, out isByRefStruct, ++depth);
+            }
+            else if (PInvokeTableGenerator.IsBlittable(t, log))
+                c = 'n';
+
+            isByRefStruct = true;
+        }
+        else
+            log.Warning("WASM0064", $"Unsupported parameter type '{t.Name}'");
```

Comment:
This is the same error code as recursion. Is that intended?

Reply:
I think this was a mistake on my part originally.